src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,90 @@
+using GraphQL.Server.Common;
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            configure(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream, long? contentLength)
+        {
+            int length;
+            byte[] jsonBytes;
+            var sharedArrayPool = ArrayPool<byte>.Shared;
+            bool wasRented = false;
+            if (contentLength.HasValue)
+            {
+                length = (int)contentLength;
```

Comment:
This logic is buggy. You'll truncate if the content is bigger than an int (that's why it's a long). This should be capped to some maximum size anyways (that is within integer range).

Reply:
👍 I was aware it'd truncate but we'd decided to let it go. Would be good to handle it though, I agree.

Would you suggest we just throw here, throw in the middleware (so we can make this interface just deal with an `int`), or can we cap it with a fx setting somehow?

Reply:
Let's introduce `int GraphQLOptions.MaxPayloadSize { get; set; }`

Reply:
```
if (contentLength != null && contentLength < 2MB) 
// Use pooled array
```
I believe 2MB is the maximum size the array pool will pool. Anything over will be heap allocated.

PS: If you want to avoid copying, you can use a `PipeReader` on the HTTP request as well to buffer up to some maximum size then read it.

Reply:
I can't seem to find any property exposing the default `Shared` array pool max size. I guess I'll hardcode it in somewhere.

Reply:
Seems a bit hard to calculate. Traced it back to [here](https://github.com/dotnet/runtime/blob/ccf6aedb63c37ea8e10e4f5b5d9d23a69bdd9489/src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs#L66). Blog post here seems to say it's 2^20, so I'll go with that hardcoded.

Reply:
yah, I read that code before commenting 😉

Reply:
I think I've sorted this part out, happy for it to be reviewed.

Reply:
> Anything over will be heap allocated.

It is true but it does not follow from this that you need to stop taking the array from the pool. This will only complicate the code. There is practically no difference in performance.
https://github.com/dotnet/runtime/blob/ccf6aedb63c37ea8e10e4f5b5d9d23a69bdd9489/src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs#L142